3.0 Code Updates

icnocop opened the pull request:
Migrated to SDK-style project
Upgraded to SpecFlow 3
Upgraded to MSTest.Framework and MSTest.Adapter
Added ability to attach files to MSTest test results context
Added ability to set the browser time zone for Microsoft Edge (Chromium) and Google Chrome
Added web driver logging support
Support Win App Driver, keeping track of the page/window/dialog history
Added TempWebElement
Added support for double-clicking, right-clicking, and mouse cursor hover
Added support for closing the browser/dialog
Added support for entering text
Added support for maximizing and switching windows, and focusing the window before clicking on elements
Added support for navigating to the previous page (going back) and refreshing the page
Added support for validating the page URL query string
Added support for setting element data using JavaScript
Added support for setting the browser page load strategy from the configuration
Added support for validating virtual properties
Added support for page sections and contexts
Added support for waiting until validation table
Improved support for elements that require custom logic to get and set values

Fixes #48
Fixes #66
Fixes #71
Fixes #118
Fixes #170
Fixes #171
Fixes #176
Fixes #183
Fixes #185
Fixes #186
Fixes #187
Fixes #192
Fixes #197

icnocop committed:
Upgraded to Visual Studio 2019
Migrated to SDK-style project
Upgraded to SpecFlow 3
Upgraded to MSTest.Framework and MSTest.Adapter
Added ability to attach files to MSTest test results context
Added ability to set the browser time zone for Microsoft Edge (Chromium) and Google Chrome
Added web driver logging support
Support Win App Driver, keeping track of the page/window/dialog history
Added TempWebElement
Added support for double-clicking, right-clicking, and mouse cursor hover
Added support for closing the browser/dialog
Added support for entering text
Added support for maximizing and switching windows, and focusing the window before clicking on elements
Added support for navigating to the previous page (going back) and refreshing the page
Added support for validating the page URL query string
Added support for setting element data using JavaScript
Added support for setting the browser page load strategy from the configuration
Added support for validating virtual properties
Added support for page sections and contexts
Added support for waiting until validation table
Improved support for elements that require custom logic to get and set values

icnocop commented:
Fixes #48
Fixes #66
Fixes #71
Fixes #118
Fixes #170
Fixes #171
Fixes #176
Fixes #183
Fixes #185
Fixes #186
Fixes #187
Fixes #192
Fixes #197

dpiessens commented:
Hi Rami,
This is a big contribution, thanks! Just as a heads up, the review on this is going to take a while. 247 files and 13 issues in a PR is a lot to mentally digest. I'll make comments as we go, so others can understand the conversation. I do think we should version this as 3.0 given the number of changes.

One larger theme that comes to mind is should we continue support of CodedUI. Based on this article: https://devblogs.microsoft.com/devops/changes-to-coded-ui-test-in-visual-studio-2019/ the framework seems to be deprecated. Thoughts?

dpiessens commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,146 @@
+  workflow_dispatch:
+
+env:
+  VERSION: 2.1.1.${{ github.run_number }}
```
Let's bump this to 3.0.0

icnocop replied:
Done. Thank you.

dpiessens commented on examples/FullDemo/ContosoUniversity.AcceptanceTests/App.config:
```diff
@@ -13,9 +13,5 @@
     </stepAssemblies>
-    <unitTestProvider name="MsTest"/>
-    <plugins>
-      <add name="SpecBindGeneratorPlugin" type="Generator"/>
```
What's the new plugin strategy here? does it follow: https://docs.specflow.org/projects/specflow/en/latest/Extend/Plugins.html

icnocop replied:
From https://docs.specflow.org/projects/specflow/en/latest/Integrations/CodedUI.html#introduction, "Coded UI is no longer supported with SpecFlow 3."

Instead of using a plugin to add attributes to the generated test class, there are [explicit calls `Playback.Initialize();` and `Playback.Cleanup();`](https://github.com/SpecBind/specbind/blob/d0b84bf9d6be99adf115a31a0eab3df82951fb5c/src/SpecBind.CodedUI/PlaybackSteps.cs#L24).

This requires adding `SpecBind.CodedUI` as a step assembly in the application configuration file.

I'm thinking we can add this to `App.config.install.xdt` so it's automatically added when installing the `SpecBind.CodedUI` nuget package in a subsequent PR.

dpiessens commented on .gitignore:
```diff
@@ -115,3 +114,4 @@ UpgradeLog*.XML
 src/SpecBind.sln.DotSettings
 examples/FullDemo/ContosoUniversity/App_Data/School.sdf
 /examples/FullDemo/ContosoUniversity/Properties/PublishProfiles/ContosoUniversity.pubxml
+*.feature.cs
```
Should feature files really be excluded from SCM?

icnocop replied:
Only the generated code-behind is being excluded.

They're automatically generated when you build the project anyway.

Otherwise, if the feature file is changed, then you will always see two changed files, when actually you just changed one.

SpocWeb replied:
I also started to exclude the generated Code. The Feature Folder looks neat now, so that 'Business' Users have fewer excuses.

dpiessens commented on src/SpecBind.CodedUI.IntegrationTests/Features/AlertTests.feature:
```diff
@@ -1,42 +1,44 @@
-	  And I see an alert box and select Ok
-	 Then I see
+﻿@mstest:DeploymentItem:TechTalk.SpecFlow.MSTest.SpecFlowPlugin.dll
+@mstest:DeploymentItem:SpecBind.MsTest.Steps.dll
```
Adding these for users will be a major breaking change. Similar to above, can we make sure that this is the correct way to implement a plugin like this?

icnocop replied:
There was an issue filed for this [here](https://github.com/SpecFlowOSS/SpecFlow/issues/2076) and I recently created a PR to resolve it [here](https://github.com/SpecFlowOSS/SpecFlow/pull/2301), at least for `TechTalk.SpecFlow.MSTest.SpecFlowPlugin.dll`.

This is only an issue when deployment is enabled in test/run settings using MSTest.
`SpecBind.MsTest.Steps.dll` is optional, and only required when you want to take advantage of attaching files to test results when using MSTest.

There may be a way to include `SpecBind.MsTest.Steps.dll` in test/run settings to avoid having to add this to feature files, and I think that's out of SpecBind's scope.

dpiessens commented on src/SpecBind.CodedUI/CodedUIBrowserFactory.cs:
```diff
@@ -56,14 +58,23 @@ protected override IBrowser CreateBrowser(ILogger logger)
+            Playback.PlaybackSettings.SmartMatchOptions = SmartMatchOptions.None; // TopLevelWindow | Control
+            Playback.PlaybackSettings.ThinkTimeMultiplier = 0; // 1
+            Playback.PlaybackSettings.WaitForReadyLevel = WaitForReadyLevel.Disabled; // UIThreadOnly
+            Playback.PlaybackSettings.WaitForReadyTimeout = (int)this.Configuration.PageLoadTimeout.TotalMilliseconds; // 60000
```
We need to document what changes this will have on users. I suspect changing some of them might cause issues.

dpiessens commented on src/SpecBind.Selenium.IntegrationTests/SpecBind.Selenium.IntegrationTests.csproj:
```diff
@@ -1,187 +1,65 @@
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
-    <DebugSymbols>true</DebugSymbols>
+    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
+    <UseWindowsForms>true</UseWindowsForms>
```
Do we need this?

icnocop replied:
Yes.

[`GenerateAssemblyInfo` has a default value of `true`](https://docs.microsoft.com/en-us/dotnet/core/migration/assembly-info#generateassemblyinfo) which would otherwise conflict with the existing AssemblyInfo.cs files and cause build failures.

For example,
```
C:\github\icnocop\specbind\src\SpecBind\obj\Debug\net472\SpecBind.AssemblyInfo.cs(14,12,14,54): error CS0579: Duplicate 'System.Reflection.AssemblyCompanyAttribute' attribute
C:\github\icnocop\specbind\src\SpecBind\obj\Debug\net472\SpecBind.AssemblyInfo.cs(15,12,15,60): error CS0579: Duplicate 'System.Reflection.AssemblyConfigurationAttribute' attribute
C:\github\icnocop\specbind\src\SpecBind\obj\Debug\net472\SpecBind.AssemblyInfo.cs(16,12,16,58): error CS0579: Duplicate 'System.Reflection.AssemblyFileVersionAttribute' attribute
C:\github\icnocop\specbind\src\SpecBind\obj\Debug\net472\SpecBind.AssemblyInfo.cs(18,12,18,54): error CS0579: Duplicate 'System.Reflection.AssemblyProductAttribute' attribute
C:\github\icnocop\specbind\src\SpecBind\obj\Debug\net472\SpecBind.AssemblyInfo.cs(19,12,19,52): error CS0579: Duplicate 'System.Reflection.AssemblyTitleAttribute' attribute
C:\github\icnocop\specbind\src\SpecBind\obj\Debug\net472\SpecBind.AssemblyInfo.cs(20,12,20,54): error CS0579: Duplicate 'System.Reflection.AssemblyVersionAttribute' attribute
```

If desired, we can try to delete the existing `AssemblyInfo.cs` files, and remove this MSBuild property, so the `AssemblyInfo.cs` files are automatically generated, in a subsequent PR.

icnocop committed:
Bumped up version to 3.0.0

icnocop commented:
It's true that Coded UI is being deprecated.
I'm thinking we can at least try to keep it around as long as it's not too hard to maintain.

I'm not expecting all features in SpecBind to be implemented in Coded UI.
For example, some features may only exist when using Selenium.

icnocop committed:
Deleted generated code

dpiessens commented on src/SpecBind.Selenium.Tests/SpecBind.Selenium.Tests.csproj:
```diff
@@ -1,132 +1,40 @@
+    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
+    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
+    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
```
Why are we generating XML docs for test classes?

dpiessens commented on src/SpecBind.Selenium/Drivers/SeleniumDriverBase.cs:
```diff
@@ -14,11 +13,12 @@ namespace SpecBind.Selenium.Drivers
     /// The base class for Selenium Drivers.
     /// </summary>
-    internal abstract class SeleniumDriverBase : ISeleniumDriver
+    public abstract class SeleniumDriverBase : ISeleniumDriver
```
Why are we making this class public? The goal with any API is to lead developers to the "pit of success", so unless we're ignoring selenium drivers entirely then a consumer of SpecBind should never need to use this class directly.

dpiessens commented on src/SpecBind.Selenium/Drivers/SeleniumEdgeDriver.cs:
```diff
@@ -4,28 +4,87 @@
+            //
+            // The host can be specified on the command line as follows:
+            // MicrosoftWebDriver.exe --host=127.0.0.1
+            edgeDriverService.Host = "127.0.0.1"; // or localhost
```
Does the behavior change with localhost? Some services are beginning to disable IPv4 on workstations so localhost would be more compatible with IPv6.

dpiessens commented on src/SpecBind.Selenium/Drivers/SeleniumEdgeDriver.cs:
```diff
@@ -4,28 +4,87 @@
-            return new EdgeDriver(edgeDriverService, edgeOptions);
+            var edgeOptions = new EdgeOptions
+            {
+                PageLoadStrategy = (PageLoadStrategy)Enum.Parse(typeof(PageLoadStrategy), browserFactoryConfiguration.PageLoadStrategy, true),
```
Nit here, I think `Enum.TryParse<PageLoadStrategy>(browserFactoryConfiguration.PageLoadStrategy, out var strategy) ? strategy : PageLoadStrategy.{Default}` would be better

dpiessens commented on src/SpecBind.Selenium/Drivers/SeleniumInternetExplorerDriver.cs:
```diff
@@ -4,24 +4,63 @@
+        {
+            base.Stop(scenarioContext);
+
+            TestResultFileNotifier testResultFileNotifier = scenarioContext.ScenarioContainer.Resolve<TestResultFileNotifier>();
```
Do we want a helper around this test file code, or put it in the base method? It seems to be the same pattern everywhere.

dpiessens commented on src/SpecBind.Selenium/Drivers/SeleniumWindowsDriver.cs:
```diff
@@ -0,0 +1,147 @@
+                    if (scenarioHelper != null)
+                    {
+                        bool isError = scenarioHelper.GetError() != null;
+                        string stepFileName = scenarioHelper.GetStepFileName(isError);
```
Is there a difference in the log for windows vs. other platforms or drivers? I think we want this log file pattern to be as consistent as possible across drivers.

dpiessens commented on src/SpecBind.Selenium/NativeMethods.cs:
```diff
@@ -0,0 +1,78 @@
+    /// <summary>
+    /// Native Methods.
+    /// </summary>
+    internal static class NativeMethods
```
I'm a little concerned with all this native process handling... SpecFlow and Seleium work cross-platform and moving outside of .NET forces us to lock into windows. What's the intent with this and the TCP stuff?

dpiessens commented on src/SpecBind.Selenium/SeleniumApplication.cs:
```diff
@@ -0,0 +1,281 @@
+        /// </value>
+        public Lazy<ISeleniumDriver> SeleniumDriver { get; }
+
+        private WindowsDriver<WindowsElement> WindowsDriver => this.Driver as WindowsDriver<WindowsElement>;
```
Is the point of this class to run on windows? What if someone wants to use Linux?

dpiessens commented on src/SpecBind.Selenium/SeleniumApplication.cs:
```diff
@@ -0,0 +1,281 @@
+        /// <returns>The cookie (if exists)</returns>
+        public override Cookie GetCookie(string name)
+        {
+            throw new NotSupportedException();
```
Are all of these really not supported, or just not needed?

dpiessens commented on src/SpecBind.Selenium/SeleniumBase.cs:
```diff
@@ -7,26 +7,29 @@ namespace SpecBind.Selenium
     public abstract class SeleniumBase : BrowserBase
     {
-        private Lazy<IWebDriver> driver;
+        private Lazy<IWebDriverEx> driver;
```
Why did we change IWebDriver to IWebDriverEx? Did it conflict with a class from IWebDriver?

dpiessens commented on src/SpecBind.Selenium/SeleniumBase.cs:
```diff
@@ -111,10 +117,10 @@ public override string TakeScreenshot(string imageFolder, string fileNameBase)
             try
             {
-                var fullPath = Path.Combine(imageFolder, $"{fileNameBase}.jpg");
+                var fullPath = Path.Combine(imageFolder, $"{fileNameBase}.png");
```
JPEGs get better compression than PNGs, was there a reason this needed to change?

dpiessens commented on src/SpecBind.Selenium/SeleniumBase.cs:
```diff
@@ -149,11 +155,109 @@ public override string SaveHtml(string destinationFolder, string fileNameBase)
+                .Replace("{F10}", Keys.F10)
+                .Replace("{F11}", Keys.F11)
+                .Replace("{F12}", Keys.F12)
+                .Replace("{NEWLINE}", Environment.NewLine);
```
You can't send a newline key in real life that I know of...

dpiessens commented on src/SpecBind.Selenium/SeleniumBase.cs:
```diff
@@ -149,11 +155,109 @@ public override string SaveHtml(string destinationFolder, string fileNameBase)
+
+            string replacedText = text
+                .Replace("{BACKSPACE}", Keys.Backspace)
+                .Replace("{BS}", Keys.Backspace)
```
Why 2 aliases for Backspace?

dpiessens commented on src/SpecBind.Selenium/SeleniumBrowser.cs:
```diff
@@ -101,7 +110,12 @@ public override Cookie GetCookie(string name)
+            {
+                IOptions options = localDriver.Manage();
+                ICookieJar cookieJar = options.Cookies;
+                cookieJar.DeleteAllCookies();
```
Two things here:
1. Why not implement this in WindowsDriverEx so it works
2. Please change the implementation line back to a single line dot setup, it's more readable.
`
localDriver.Manage().Cookies.DeleteAllCookies();
`

dpiessens commented on src/SpecBind.Selenium/SeleniumListElementWrapper.cs:
```diff
@@ -37,8 +37,15 @@ public SeleniumListElementWrapper(TElement parentElement, IBrowser browser)
+        /// Gets or sets a value indicating whether to cache the element lookup.
+        /// </summary>
+        /// <value><c>true</c> if cache; otherwise, <c>false</c>.</value>
+        public bool Cache { get; set; }
```
It been a while since I worked with the list code, but I recall a lot of issues when elements are cached. I'm assuming this has been tested to ensure there are no unexpected side effects.

dpiessens commented on src/SpecBind.Selenium/SeleniumPage.cs:
```diff
@@ -401,11 +545,17 @@ private static bool CheckElementState(Func<IWebElement, bool> checkFunc, IWebEle
+            if (element is IDataControl)
             {
-                dataControlElement.SetValue(data);
+                IDataControl dataControlElement = element as IDataControl;
```
Why not just keep the type cast variable?

dpiessens commented on src/SpecBind.Selenium/TempWebElement.cs:
```diff
@@ -0,0 +1,25 @@
+    using OpenQA.Selenium;
+
+    /// <summary>
+    /// Temp Web Element.
```
Can we add a comment about what makes this class different? Assuming it has to do with caching

dpiessens commented on src/SpecBind.Selenium/WebElement.cs:
```diff
@@ -62,11 +63,24 @@ public ICoordinates Coordinates
         /// of having to parse an element's "style" attribute to determine
         /// visibility of an element.</remarks>
-        public bool Displayed
+        public virtual bool Displayed
```
I'm a little nervous about this being virtual. There's already a lot of indirection in this framework, what behavior would need to change here?

dpiessens commented on src/SpecBind.Selenium/WebElement.cs:
```diff
@@ -302,38 +335,128 @@ public override int GetHashCode()
+                .Replace("{F10}", Keys.F10)
+                .Replace("{F11}", Keys.F11)
+                .Replace("{F12}", Keys.F12)
+                .Replace("{NEWLINE}", Environment.NewLine);
```
Does this need to be a helper function? The same code exists in the other location.

dpiessens commented on src/SpecBind.Tests/PageBaseFixture.cs:
```diff
@@ -4,8 +4,10 @@
 
 namespace SpecBind.Tests
 {
-    using System.Linq;
```
System references should be first, that's .NET defined  convention.

dpiessens commented on src/SpecBind.Tests/WebDriverSupportFixture.cs:
```diff
@@ -144,9 +150,10 @@ public void TestCheckForScreenshotWithNoErrorButSettingEnabledTakesSuccessfulScr
             WebDriverSupport.SetBrowserFactory(browserFactory.Object);
-            WebDriverSupport.SetCurrentBrowser(browser.Object);
-            var driverSupport = new WebDriverSupport(container.Object);
+            WebDriverSupport.CurrentBrowser = browser.Object;
```
FYI on this: there's a long standing design tradition in .NET about properties vs. methods. While by convention Properties should have { get; set; } instead of SetProperty(item), the assumption is that the setter has no side effects on the class. The reason I used a set method is because it does a bunch of initialization that a consumer may not know about.

dpiessens commented on src/SpecBind/ActionPipeline/PageHistoryException.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <param name="propertyName">Name of the property.</param>
+        /// <param name="pageHistory">The page history.</param>
+        public PageHistoryException(string propertyName, Dictionary<Type, IPage> pageHistory)
+                    : base(string.Join(Environment.NewLine, new[]
```
Can you make the multi-line span a static method in the class? that way you don't need the warning disable.

dpiessens commented on src/SpecBind/Actions/NavigationPostAction.cs:
```diff
@@ -5,15 +5,27 @@
     /// <summary>
     /// A post-action base class that only triggers when navigation is successful.
     /// </summary>
-    public abstract class NavigationPostAction : IPostAction
```
Did you remove extension of this class because we didn't have a use for it?

dpiessens commented on src/SpecBind/BrowserSupport/BrowserBase.cs:
```diff
@@ -110,7 +125,13 @@ public bool IsDisposed
-            this.Close();
+            try
+            {
+                this.Close();
```
Are we sure close can't cause any errors? I thought I was able to kill this when a dialog was open or I terminated the window. I just want to make sure this doesn't throw during disposal.

dpiessens commented on src/SpecBind/DataSteps.cs:
```diff
@@ -143,11 +148,56 @@ public void WhenIClearDataInFieldsStep(Table data)
+        [Given(@"I released ""(.*)""")]
+        public void GivenIReleased(string text)
+        {
+            var page = this.GetPageFromContext();
```
Is this always just a precondition, or could it also be a When?

dpiessens commented on src/SpecBind/DataSteps.cs:
```diff
@@ -143,11 +148,56 @@ public void WhenIClearDataInFieldsStep(Table data)
+        /// Given I pressed "text".
+        /// </summary>
+        /// <param name="text">The text.</param>
+        [Given(@"I pressed ""(.*)""")]
```
Is this always just a precondition, or could it also be a When?

dpiessens commented on src/SpecBind/DataSteps.cs:
```diff
@@ -143,11 +148,56 @@ public void WhenIClearDataInFieldsStep(Table data)
         /// </summary>
         /// <param name="data">The field data.</param>
         [Given(GivenObserveDataStepRegex)]
+        [When(ObserveDataStepRegex)]
```
I disagree this should be a when step. Data observation should either be a precondition or assertion. Think of scenarios like Unit tests with the Arrange, Act, Assert pattern. If you need to look at data during When then your test is either too big or testing too much.

dpiessens commented on src/SpecBind/Helpers/Logger.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// Logger.
+    /// </summary>
+    public static class Logger
```
Where specifically do we need a static class? We've gone to a lot of work to have it inject loggers through constructors even with additions of them from you. This feels like a pattern leak somehow...

dpiessens commented on src/SpecBind/Helpers/ReflectionHelper.cs:
```diff
@@ -0,0 +1,52 @@
+        /// <param name="obj">The object.</param>
+        /// <param name="propertyName">Name of the property.</param>
+        /// <returns>The value of the property.</returns>
+        public static object GetPropertyValue(object obj, string propertyName)
```
Where are we reflecting properites?

icnocop replied:
Thank you. I removed this file.

dpiessens commented on src/SpecBind/Pages/ElementLocatorAttribute.cs:
```diff
@@ -65,7 +59,7 @@ public ElementLocatorAttribute()
         /// The name.
         /// </value>
-        public string Name { get; set; }
+        public virtual string Name { get; set; }
```
I'm not sure attributes should have virtual properties, why do we need them?

icnocop replied:
This allows me to create a custom `ElementLocatorAttribute` and at least override the `Name` property `getter` so the value is dynamic.

For example, I have use cases where elements on the UI change based on the user's language, or if the UI is custom branded, or if the user is running a trial version.

With a dynamic `ElementLocatorAttribute`, I can have one feature file, and one page class which can run through these different scenarios.

Otherwise I would have to use a different feature file (or Gherikin scenario), or more complicated steps, or more complicated page class, or use a different page class for example.

dpiessens commented on src/SpecBind/Pages/PageNavigationAttribute.cs:
```diff
@@ -39,7 +39,7 @@ public PageNavigationAttribute(string url)
         /// The URL.
         /// </value>
-        public string Url { get; private set; }
+        public virtual string Url { get; private set; }
```
Same question about virtual props here

icnocop replied:
This has the same reason as my previous comment for the `Name` property.

When testing with Windows desktop applications, the `Url` is used for the window's title.

dpiessens commented on src/SpecBind/SelectionSteps.cs:
```diff
@@ -19,44 +18,81 @@ public class SelectionSteps : PageStepBase
         /// <param name="linkName">Name of the link.</param>
         [Given(GivenChooseALinkStepRegex)]
         [When(ChooseALinkStepRegex)]
+        [Then(ChooseALinkStepRegex)]
```
Same AAA Pattern I don't think choosing links should be part of validation.

icnocop replied:
Thank you. I removed this attribute.

dpiessens commented on src/SpecBind/SelectionSteps.cs:
```diff
@@ -116,5 +152,38 @@ public void GoToListItemWithCriteriaStep(string listName, Table criteriaTable)
             this.UpdatePageContext(item);
         }
+
+        private void ChooseALinkStep(string propertyName)
```
Why are these all moved into private methods? I don't see them used in multiple spots.

icnocop replied:
Thank you.

I consolidated multiple click action methods.

dpiessens reviewed (changes requested):
Thanks again for all the hard work adding this to the project. There are a decent number of comments in here, a few nits but mostly design based ones. A few high level comments:

- I'm still a bit concerned with the size of this PR overall, when I've added actions it makes it easy to check off all the boxes at the step, action, driver, what each platform implements and docs when you can see it. We need at a minimum to change the PR name to '3.0 Code Updates' and then document them all in the description alongside the issue.
- Since docs are now embedded in the code base we should have documentation updates

- The whole application concept needs to be covered and we should think about wither it needs to be in a separate assembly. Long term I see this running on .NET core and some of the interop code makes that challenging.
- Please make sure that any new actions, classes etc. added have unit tests covering them as close to 100% as possible. With few maintainers, high test coverage is the only way we can refactor without the worry that we broke something.
- I appreciate the switch to GH actions, I do want to make sure we aren't loosing any steps along the way. I haven't had time to do a full build comparison yet, but I'd like to.

icnocop committed:
Removed unused code

icnocop committed:
Removed "Then I choose (.+)" step to follow AAA pattern

icnocop committed:
Consolidated click action methods

icnocop committed:
Using codecov github workflow action

icnocop committed:
Merge pull request #1 from icnocop/coverage

Using codecov github workflow action

icnocop committed:
Using just-the-docs remote theme for github pages

icnocop commented:
Thank you for the review.

>     * I'm still a bit concerned with the size of this PR overall, when I've added actions it makes it easy to check off all the boxes at the step, action, driver, what each platform implements and docs when you can see it. We need at a minimum to change the PR name to '3.0 Code Updates' and then document them all in the description alongside the issue.

I changed the name of the PR to "3.0 Code Updates" and will work on the documentation alongside additional tests.

>     * Since docs are now embedded in the code base we should have documentation updates

I updated the GitHub pages to use a customized version of the `just-the-docs` jekyll theme for easier navigation and will will work on adding more documentation.
 
>     * The whole application concept needs to be covered and we should think about wither it needs to be in a separate assembly. Long term I see this running on .NET core and some of the interop code makes that challenging.

.NET Core support seems like a feature outside the scope of this PR.
I'm not sure if the changes I've made are the only things that would prevent us from migrating to .NET Core.

Creating a separate assembly which adds support for WinAppDriver also seems like a feature outside the scope of this PR.
`SpecBind.Selenium` isn't currently designed to support initializing Selenium drivers from external assemblies.
I think this can be addressed in a subsequent PR.

>     * Please make sure that any new actions, classes etc. added have unit tests covering them as close to 100% as possible. With few maintainers, high test coverage is the only way we can refactor without the worry that we broke something.

I can work on improving code coverage by adding tests for the code that was added and changed, as well as adding documentation for the new features.

The coverage is 81.52% right now. It was previously 84.43%.
![image](https://user-images.githubusercontent.com/507649/109410839-bb34a200-7952-11eb-82de-9c2b7867287a.png)


>     * I appreciate the switch to GH actions, I do want to make sure we aren't loosing any steps along the way. I haven't had time to do a full build comparison yet, but I'd like to.

Appveyor's environments don't have Visual Studio Enterprise installed, whereas GitHub does.
This allows us to use the official Coded UI dependencies on the build agent's machine instead of the unofficial Coded UI NuGet package.

The only other thing I noticed is Appveyor.yml sends a notification to gitter.im.
This isn't included in the GitHub workflow and I don't think it's a show stopper.
There hasn't been any activity in https://gitter.im/dpiessens/specbind since July of 2018.
I'm not sure if that space should eventually be moved to https://gitter.im/specbind/specbind, and/or GitHub Discussions enabled for this repo.

dpiessens commented:
Thanks for the updates to date. To focus the discussion down to a few areas:

- I'm not sure a discussion forum is needed at this point, so we can ignore that. I'm fine with the GH actions setup, and I will need to move package release processes over to complete that migration.

- Thanks to the updates to themes and docs, given your keen eye to complete documentation in the past I figured that would be an important part of the new actions added.

- In terms of code coverage, I'm less concerned about the percentage want to focus more what's covered.  The steps, actions and top level driver things are the ones I always tried to get as close to 100% as possible as it's small wire-up features there that are the easiest to break and having those tests give confidence when we need to refactor. The lower level interactions are more difficult and if you're confident they work then we should add an  `[ExcludeFromCoverage]` attribute explaining why. I realize I haven't expressed these goals in writing prior, so I'll be adding a contribution guide for this.

- I agree that moving to .NET core is a separate PR. I don't want to add additional work to that goal, by taking dependencies on the registry and native methods that might make that difficult. If you're against splitting out the native method portions due to the work involved, then I want to label the windows interaction as a "beta" feature and subject to breakage as we move to .NET core.

skitscha commented:
Hi, what is left to do on this pull request? There are some features that were added that I would like to leverage.

icnocop commented:
I had some difficulty trying to isolate the code paths which require test coverage when developing locally to satisfy bullet point #3.

The feedback loop was taking a little too long.

Do you have any advice or recommendations on a better approach which can quickly identify the code paths that need test coverage?

Thank you.
